integration_tests/tests/integration/in_cluster.rs

```diff
@@ -333,3 +333,244 @@ async fn in_cluster_fixed_size() {
+    use janus_core::{test_util::install_test_trace_subscriber, vdaf::VdafInstance};
+    use janus_messages::{AggregationJobId, CollectionJobId, TaskId};
+    use rand::random;
+    use reqwest::StatusCode;
+    use serde::Deserialize;
+    use std::{env, fs::File};
+    use url::Url;
+
+    /// Configuration for the rate limit test. We need to know the QPS and the window over which
+    /// it is enforced so that we can send the appropriate number of requests. We load this config
+    /// from a file so that an integration test harness that knows what rate limits have been
+    /// deployed to the cluster we test against can set the values.
+    #[derive(Deserialize)]
+    struct TestConfig {
+        window: u64,
+        upload_qps: u64,
+        aggregation_job_qps: u64,
+        collection_job_qps: u64,
+    }
+
+    impl TestConfig {
+        fn load() -> Self {
+            serde_json::from_reader(
+                File::open(
+                    env::var("JANUS_E2E_RATE_LIMIT_TEST_CONFIG")
+                        .unwrap()
+                        .as_str(),
+                )
+                .unwrap(),
+            )
+            .unwrap()
+        }
+    }
+
+    struct RequestUrlParameters<'a> {
+        leader_url: &'a Url,
+        helper_url: &'a Url,
+        first_task_id: &'a TaskId,
+        second_task_id: &'a TaskId,
+    }
+
+    async fn run_rate_limit_test(
+        request_url_maker: &dyn Fn(&RequestUrlParameters) -> (Url, Url),
+        rate_limit_picker: &dyn Fn(&TestConfig) -> u64,
+        method: Method,
+    ) {
+        install_test_trace_subscriber();
+        let test_config = TestConfig::load();
+
+        let janus_pair =
+            InClusterJanusPair::new(VdafInstance::Prio3Count, QueryType::TimeInterval).await;
+
+        let (leader_url, helper_url) = janus_pair
+            .task_parameters
+            .endpoint_fragments
+            .endpoints_for_host_client(janus_pair.leader.port(), janus_pair.helper.port());
+        let other_task_id = random();
+
+        // Send requests to two different tasks, to prove that rate limits are per-task ID. It
+        // doesn't matter that one of the tasks doesn't exist for purposes of rate limits.
+        let (first_request_url, second_request_url) = request_url_maker(&RequestUrlParameters {
+            leader_url: &leader_url,
+            helper_url: &helper_url,
+            first_task_id: &janus_pair.task_parameters.task_id,
+            second_task_id: &other_task_id,
+        });
+        let rate_limit = rate_limit_picker(&test_config);
+        let rate_limit_excess = 0.1;
+        let total_requests_count =
+            ((1.0 + rate_limit_excess) * (rate_limit * test_config.window) as f64) as u64;
+
+        let mut acceptable_status_count = 0;
+        let mut too_many_requests_count = 0;
+
+        // Send ten requests at a time, because otherwise the kubectl port-forward gets overwhelmed
+        for _ in 0..(total_requests_count / 10) {
```

Comment:
This block works, but it's kind of awkward: we generate groups of 10 requests at a time, execute them concurrently, then move on to the next group of 10. Other than running an incorrect number of requests if the `total_request_count` is not a multiple of 10, this also unnecessarily splits processing of each request into two pieces & otherwise requires slightly more complicated code. (It also reduces throughput somewhat by introducing more waiting on requests than necessary, but that is probably extremely marginal given that this is a test.)

I suggest instead introducing a [Semaphore](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html) with 10 permits. Have each spawned async task acquire a permit before making its request, and release the permit once the response is received. Spawn all tasks at once, and have each task be responsible for updating `acceptable_status_count` & `too_many_requests_count` (either by updating them atomically, hiding access behind a mutex, or just by returning a value from each async task indicating which variable to update).

(An alternative to using a semaphore would be to start 10 worker tasks, then have them execute requests until enough requests have been executed, synchronizing on a counter or perhaps a queue. This would scale a bit better than one-task-per-request but the additional complexity probably isn't justified in a test.)

Reply:
This is much better with a semaphore, thank you for that suggestion. I considered updating `acceptable_status_count` etc. the way you suggested, but we need the test to block until all the spawned tasks are done anyway, meaning we'd have to `await` all the handles to get at least a `()` value, so getting a `(HeaderValue, HttpStatus)` as we currently do doesn't seem any worse.